### DISCUSSION TITLE
Construct `Row` manually from Tuple

### DISCUSSION THREAD

**Discussion by luciengaitskell** [sqlalchemy/sqlalchemy#8002] [Usage Questions] [answered]
Is it possible to manually construct a `Row` object from a tuple and a list of columns (e.g. `Select.selected_columns`)? I'd like to leverage the built in decoding in SQLAlchemy for the types described by the `Select` object with data received from an external engine.

I'm using the [SnowflakeIO Beam transform](https://beam.apache.org/documentation/io/built-in/snowflake/#using-snowflakeio-in-python-sdk) to run a query. I compile my SQLA statement and pass the query in manually, then the data is received back just as tuples.

**Comment by zzzeek** [comment]
why not use the [actual snowflake dialect](https://github.com/snowflakedb/snowflake-sqlalchemy) ?

**Reply by zzzeek** [reply]
unless this is something different.

**Reply by luciengaitskell** [reply]
SnowflakeIO is more optimized for Beam as it runs the query, reads into a CSV file, then is received row-by-row by the workers in Python.

**Comment by zzzeek** [comment, marked as answer]
anyway, you wouldn't create Row objects you'd create Result objects.  here is the simplest example

```py
from sqlalchemy.engine.result import IteratorResult
from sqlalchemy.engine.result import SimpleResultMetaData


row_metadata = SimpleResultMetaData(
    ['a', 'b','c']
)

result = IteratorResult(
    row_metadata,
    iter([(1, 2, 3), (4, 5, 6)]),
)

for row in result:
    print(row._fields)
    print(row)

```

**Reply by luciengaitskell** [reply]
Thanks for the example. Could I use a list of `Column`s when building a `SimpleResultMetaData` object?

**Reply by zzzeek** [reply]
there's not public API for that right now but they can be added using private state on the object.

**Reply by luciengaitskell** [reply]
Would `Result.columns(...)` work in this case?

Edit: I see this just selects columns to return

**Reply by luciengaitskell** [reply]
@zzzeek Just to confirm, there's no mechanism to manually specify a tuple with columns and decode them?

**Reply by zzzeek** [reply]
what does "decode" mean ?

**Reply by zzzeek** [reply]
as i mentioned previously, there's private state on the object where you can add columns.  but i can't guarantee that API wont break even though it is unlikely in the near term.

(this is where you see if that private API can work and then you work on a PR to create fully tested public API for said elements)

**Reply by luciengaitskell** [reply]
> what does "decode" mean ?

I'm looking to use the built-in column type parsing manually. Could I run this operation manually? For example if I have a raw value (likely string), can I pass it to a Column type or a 'type' type and parse it into the proper Python native object.

This could be solved by using the `process_result_value` method for each type, but there might be a higher level operation?

**Reply by zzzeek** [reply]
oh you want to send in the processors.  that's much easier.     these can be sent as `_processors`:

```py
from sqlalchemy import DateTime
from sqlalchemy import Numeric
from sqlalchemy import String
from sqlalchemy.dialects import sqlite
from sqlalchemy.engine.result import IteratorResult
from sqlalchemy.engine.result import SimpleResultMetaData

dialect = sqlite.dialect()

row_metadata = SimpleResultMetaData(
    ["a", "b", "c"],
    _processors=(
        String().dialect_impl(dialect).result_processor(dialect, None),
        DateTime().dialect_impl(dialect).result_processor(dialect, None),
        Numeric().dialect_impl(dialect).result_processor(dialect, None),
    ),
)

result = IteratorResult(
    row_metadata,
    iter(
        [
            ("string1", "2022-03-05 12:15:17", 56.238490),
            ("string1", "2022-09-10 12:15:17", 0.34534),
        ]
    ),
)

for row in result:
    print(row._fields)
    print(row)

```


however, you don't really need to use `_processors` you could just process the rows as you put them in also, so no private API needed:

```py
# same example no private API

from sqlalchemy import DateTime
from sqlalchemy import Numeric
from sqlalchemy import String
from sqlalchemy.dialects import sqlite
from sqlalchemy.engine.result import IteratorResult
from sqlalchemy.engine.result import SimpleResultMetaData

dialect = sqlite.dialect()

processors = (
    String().dialect_impl(dialect).result_processor(dialect, None),
    DateTime().dialect_impl(dialect).result_processor(dialect, None),
    Numeric().dialect_impl(dialect).result_processor(dialect, None),
)

row_metadata = SimpleResultMetaData(
    ["a", "b", "c"],
)

result = IteratorResult(
    row_metadata,
    (
        tuple(
            proc(elem) if proc is not None else elem
            for proc, elem in zip(processors, row)
        )
        for row in [
            ("string1", "2022-03-05 12:15:17", 56.238490),
            ("string1", "2022-09-10 12:15:17", 0.34534),
        ]
    ),
)

for row in result:
    print(row._fields)
    print(row)

```

we can see in the output the strings/floats are converted to datetime/ Decimal:

```
$ python test3.py 
('a', 'b', 'c')
('string1', datetime.datetime(2022, 3, 5, 12, 15, 17), Decimal('56.2384900000'))
('a', 'b', 'c')
('string1', datetime.datetime(2022, 9, 10, 12, 15, 17), Decimal('0.3453400000'))
```

**Reply by luciengaitskell** [reply]
Thanks for the example. That works well on `sqlite`, but I've discovered the `snowflake` dialect doesn't appear to implement the type processors in the same way.

It seems like I was misunderstanding how the result processors work, as I thought I could use them to cast a string representation to a builtin type (int, datetime, etc.) using SQLA types. It's looking like I'll have to manually define these manually instead.

**Comment by CaselIT** [comment]
@zzzeek maybe we could create a faq item with your example?

Like "How do I create mock results, for example for unit tests"?

**Reply by zzzeek** [reply]
That's a very specific pattern they are looking for above and the issue of mocking results is way bigger than that in the general sense.  

I do think we could have SimpleResultMetaData in the docs, it's public yet not included in the documentation right now.  My example above could be in the docstring for SimpleResultMetaData.

**Reply by CaselIT** [reply]
makes sense to document.

I've proposed that faq, since there was another question regarding it in the chat, and it's something that may be useful in some cases in the tests, so maybe we could document and in the faq we mention that "for very simple cases" SimpleResultMetaData  may be used

**Reply by zzzeek** [reply]
fire away

I dont know how often people are even finding the FAQ these days
